src/staking-v3/logic/services/DappStakingService.ts

```diff
@@ -0,0 +1,684 @@
+
+    return await this.dappStakingRepository.getStakerInfo(address, includePreviousPeriods);
+  }
+
+  public async startAccountLedgerSubscription(address: string): Promise<void> {
+    Guard.ThrowIfUndefined('address', address);
+
+    await this.dappStakingRepository.startAccountLedgerSubscription(address);
+  }
+
+  private async getStakerEraRange(senderAddress: string) {
+    const [protocolState, ledger, constants] = await Promise.all([
+      this.dappStakingRepository.getProtocolState(),
+      this.dappStakingRepository.getAccountLedger(senderAddress),
+      await this.dappStakingRepository.getConstants(),
+    ]);
+    let rewardsExpired = false;
+
+    // *** 1. Determine last claimable era.
+    const currentPeriod = protocolState!.periodInfo.number;
+    const firstStakedEra = Math.min(
+      ledger.staked.era > 0 ? ledger.staked.era : Infinity,
+      ledger.stakedFuture?.era ?? Infinity
+    );
+    const lastStakedPeriod = Math.max(ledger.staked.period, ledger.stakedFuture?.period ?? 0);
+    let lastStakedEra = 0;
+
+    if (
+      this.hasRewardsExpired(lastStakedPeriod, currentPeriod, constants.rewardRetentionInPeriods)
+    ) {
+      // Rewards expired.
+      rewardsExpired = true;
+    } else if (lastStakedPeriod < currentPeriod) {
+      // Find last era from past period.
+      const periodInfo = await this.dappStakingRepository.getPeriodEndInfo(lastStakedPeriod);
+      lastStakedEra = periodInfo?.finalEra ?? 0; // periodInfo shouldn't be undefined for this case.
+    } else if (lastStakedPeriod === currentPeriod) {
+      // Find last era from current period.
+      lastStakedEra = protocolState!.era - 1;
+    } else {
+      throw 'Invalid operation.';
+    }
+
+    if (firstStakedEra > lastStakedEra) {
+      // No rewards earned. See if we need to distinguish this and rewards expired.
+      rewardsExpired = true;
+    }
+
+    const firstSpanIndex = firstStakedEra - (firstStakedEra % constants.eraRewardSpanLength);
+    const lastSpanIndex = lastStakedEra - (lastStakedEra % constants.eraRewardSpanLength);
+
+    return {
+      firstStakedEra,
+      lastStakedEra,
+      firstSpanIndex,
+      lastSpanIndex,
+      rewardsExpired,
+      eraRewardSpanLength: constants.eraRewardSpanLength,
+    };
+  }
+
+  private hasRewardsExpired(
+    stakedPeriod: number,
+    currentPeriod: number,
+    rewardRetentionInPeriods: number
+  ): boolean {
+    return stakedPeriod < currentPeriod - rewardRetentionInPeriods;
+  }
+
+  private async signCall(
+    call: ExtrinsicPayload,
+    senderAddress: string,
+    successMessage: string
+  ): Promise<void> {
+    const wallet = this.walletFactory();
+    await wallet.signAndSend({
+      extrinsic: call,
+      senderAddress: senderAddress,
+      successMessage,
+    });
+  }
+
+  private async getMultiplier(): Promise<bigint> {
+    const metadata = await this.metadataRepository.getChainMetadata();
+    return BigInt(10 ** metadata.decimals);
+  }
```

Comment:
Dumb question - I noticed this being used in the code but don't understand why it's needed. Could you please explain?

Reply:
This is because I didn't find a better way to divide two BigInt numbers. Lets say you have two BigInt numbers `a` and `b` and `a<b`, than `a / b = 0`.
I modified formula a bit `(a * multiplier / b) / multiplier`

Reply:
Ok, I see!

So for example this:
```
 (((info.staked.voting * multiplier) / periodEndInfo.totalVpStake) *
              periodEndInfo.bonusRewardPool) /
            multiplier;
```

is essentially: `info.staked.voting * multiplier / periodEndInfo.totalVpStake * periodEndInifo.bonusRewardPool * multiplier`

To make it more in line with how backend works, I'd suggest:
`info.staked.voting * periodEndInfo.bonusRewardPool / periodEndInfo.totalVpStake`

This isn't 100% accurate as runtime does it, but it's very close.
And this way you don't need the multiplier.

Reply:
Fixed, thank you